app/src/main/java/com/example/drawingactivity/DialogUtils.kt

```diff
@@ -0,0 +1,151 @@
+                        viewModel.addLayer(layerName)
+                    }
+                }
+                viewModel.dialogCompleted()
+            }
+            setNegativeButton(context.getString(R.string.addLayerNegativeButton)) { _, _ ->
+                viewModel.dialogCompleted()
+            }
+            setOnCancelListener {
+                viewModel.dialogCompleted()
+            }
+        }
+        dialog.show()
+    }
+
+    fun editLayerDialog(index: Int) {
+        viewModel.isEditDialogOpen = true
+        val inflater = LayoutInflater.from(context)
+        val view = inflater.inflate(R.layout.add_layer_dialog, null)
+        val dialog = AlertDialog.Builder(context).apply {
+            setTitle(context.getString(R.string.editLayerDialogTitle))
+            setView(view)
+            val errorView = view.findViewById<TextView>(R.id.errorText)
+            val editView = view.findViewById<EditText>(R.id.layerNameField)
+            while (!viewModel.layerHintFound) {
+                val hint = context.getString(R.string.layerHint, viewModel.hintIndex)
+                viewModel.createLayerHint(hint)
+            }
+            editView.hint = context.getString(R.string.layerHint, viewModel.hintIndex)
+            if (viewModel.isEditDialogOpen && viewModel.userNewLayerName != null) {
+                editView.setText(viewModel.userNewLayerName)
+                errorView.isVisible = editView.text.toString().length > DrawViewModel.CHAR_LIMIT
+            }
+            editView.addTextChangedListener {
+                viewModel.userNewLayerName = it.toString().trim()
+                errorView.isVisible =
+                    editView.text.toString().trim().length > DrawViewModel.CHAR_LIMIT
+            }
+            viewModel.editLayerIndex = index
+            setPositiveButton(context.getString(R.string.addLayerPositiveButton)) { _, _ ->
+                var layerName: String = editView.text.toString().trim()
+                when (layerName) {
+                    context.getString(R.string.layerHint, viewModel.hintIndex) -> {
+                        viewModel.editLayer(layerName, index)
+                    }
+                    DrawViewModel.NAME_EMPTY -> {
+                        layerName = context.getString(R.string.layerHint, viewModel.hintIndex)
+                        viewModel.editLayer(layerName, index)
+                    }
+                    else -> {
+                        viewModel.editLayer(layerName, index)
+                    }
+                }
+                viewModel.dialogCompleted()
+                layerListBottomSheet()
+            }
+            setNegativeButton(context.getString(R.string.addLayerNegativeButton)) { _, _ ->
+                viewModel.dialogCompleted()
+            }
+            setOnCancelListener {
+                viewModel.dialogCompleted()
+            }
+        }
+        dialog.show()
+    }
+
+    fun layerListBottomSheet() {
```

Comment:
should ideally have a way to tell what index you were looking at for when the user was editing and they get a screen interruption like rotation. the viewState stuff from the other day will help with this.